### PR TITLE
bpo-37596: Make `set` and `frozenset` marshalling deterministic

### PR DESCRIPTION
**Pull request by brandtbucher**
This PR is intended as an alternative to #27769.

https://bugs.python.org/issue37596

### CODE REVIEW

**Comment by brandtbucher** [comment]
Buildbot failure looks unrelated (it seems to have been having that issue for several days).

File: Lib/test/test_marshal.py [hunk 1]
```diff
@@ -344,6 +344,20 @@ def test_eof(self):
         for i in range(len(data)):
             self.assertRaises(EOFError, marshal.loads, data[0: i])
 
+    def test_deterministic_sets(self):
+        # bpo-37596: sets and frozensets must always marshal deterministically!
```
Comment (rhettinger) [inline Lib/test/test_marshal.py]: The word "deterministic" is correct but doesn't help motivate the algorithm to someone who hasn't read the tracker issue.  I recommend:

> To support reproducible builds, sets and frozensets need to be stored with the keys in consistent order even when they have been scrambled by hash randomization.  Here we use an order equivalent to:  sorted(sorted(s, key=marshal.dumps)).  See bpo-37596 for details.

File: Lib/test/test_marshal.py [hunk 2]
```diff
@@ -344,6 +344,20 @@ def test_eof(self):
...
+        # Test this by seeding string hashes:
+        elements = "(float('nan'), b'a', b'b', b'c', 'x', 'y', 'z')"
+        for kind in ("set", "frozenset"):
+            script = f"import marshal; print(marshal.dumps({kind}({elements})))"
+            with self.subTest(kind):
+                # {nan, b'b', 'x', b'a', b'c', 'y', 'z'}
```
Comment (rhettinger) [inline Lib/test/test_marshal.py]: I suggest adding Serhiy's test case as well, ``{("string", 1), ("string", 2), ("string", 3)}``.  If someone later tries to "simplify" this algorithm, the test would provide early detection of the backreference problem.

File: Misc/NEWS.d/next/Library/2021-08-23-21-39-59.bpo-37596.ojRcwB.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Ensure that :class:`set` and :class:`frozenset` objects are always
+:mod:`marshalled <marshal>` deterministically.
```
Comment (rhettinger) [inline Misc/NEWS.d/next/Library/2021-08-23-21-39-59.bpo-37596.ojRcwB.rst]: s/deterministically/reproducibly/

File: Python/marshal.c [hunk 4]
```diff
@@ -503,9 +503,41 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
             W_TYPE(TYPE_SET, p);
         n = PySet_GET_SIZE(v);
         W_SIZE(n, p);
+        // bpo-37596: We need to write the elements in a deterministic order!
```
Comment (rhettinger) [inline Python/marshal.c]: Try the comment wording suggested above:

The word "deterministic" is correct but doesn't help motivate the algorithm to someone who hasn't read the tracker issue. I recommend:

> To support reproducible builds, sets and frozensets need to be stored with the keys in consistent order even when they have been scrambled by hash randomization. Here we use an order equivalent to: sorted(sorted(s, key=marshal.dumps)). See bpo-37596 for details.

File: Python/marshal.c [hunk 5]
```diff
@@ -503,9 +503,41 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
...
+        }
         while (_PySet_NextEntry(v, &pos, &value, &hash)) {
+            PyObject *dump = PyMarshal_WriteObjectToString(value, p->version);
+            if (dump == NULL) {
+                p->error = WFERR_UNMARSHALLABLE;
+                goto anyset_done;
+            }
+            PyObject *pair = PyTuple_Pack(2, dump, value);
+            Py_DECREF(dump);
+            int error = pair == NULL || PyList_Append(pairs, pair);
```
Comment (rhettinger) [inline Python/marshal.c]: This line is a bit awkward and hard to read.

File: Python/marshal.c [hunk 6]
```diff
@@ -503,9 +503,41 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
...
+            Py_XDECREF(pair);
+            if (error) {
+                p->error = WFERR_NOMEMORY;
+                goto anyset_done;
+            }
+        }
+        if (PyList_Sort(pairs)) {
+            p->error = WFERR_UNMARSHALLABLE;
```
Comment (rhettinger) [inline Python/marshal.c]: What could trigger this error?  I'm thinking that the only failure possible at this point would be *WFERR_NOMEMORY*

**Review by rhettinger — APPROVED** [review]

**Review by rhettinger — APPROVED** [review]

**Pull request merged**